### PR TITLE
persist DHT to json file on disk

### PR DESCRIPTION
**Pull request by xaiki**
this sets the groundwork for the webtorrent persist DHT feature

### CODE REVIEW

File: client.js [hunk 1]
```diff
@@ -30,6 +31,9 @@ var MAX_CONCURRENCY = 3 // α from Kademlia paper
 var ROTATE_INTERVAL = 5 * 60 * 1000 // rotate secrets every 5 minutes
 var SECRET_ENTROPY = 160 // entropy of token secrets
 var SEND_TIMEOUT = 2000
+var HOMEDIR = (process.platform === 'win32') ? process.env.HOMEPATH : process.env.HOME
+var PERSIST_DHT_FILE = HOMEDIR + '/.bitorrent-dht.persist.json'
```
Comment (feross) [inline client.js]: This should use `path.join` instead of a `/`.

File: client.js [hunk 2]
```diff
@@ -131,19 +137,45 @@ function DHT (opts) {
...
 
   process.nextTick(function () {
+    var persistedNodes = []
+    if (opts.persist !== false) {
+      try {
+        persistedNodes = require(self.persistFile)
+        if (!Array.isArray(persistedNodes)) {
+          persistedNodes = []
+          debug('error reading ' + self.persistFile + ': returned object is not an array')
+        }
```
Comment (feross) [inline client.js]: This `if` block should be moved out of the `try-catch`. The only part we *expect* to throw an exception is the call to `require`, so everything else should live outside the `try-catch`.

**Comment by feross** [comment]
My gut tells me that this logic belongs in `webtorrent`, i.e. the user of `bittorrent-dht` should handle it. But I'll think about it a bit more. @mafintosh - do you have any thoughts?

**Comment by xaiki** [comment]
I'll get to your suggestions tomorrow.

But I'll argue that this is dht behaviour and hence belongs here.

You do need to hook before bootstrap (as you need to control it) and putting in another layer would make this weird (as the layer should tell dht to bootstrap, when, how... You can see the unelegancy of this)
What is lacking is a localstorage fallback for the browser case, (or does require('fs') magically work)

Cheers



—
Sent from Mobile,
I shalt be brief

**Comment by mafintosh** [comment]
If we wanna add persistence I think we should support passing in a [levelup](https://github.com/rvagg/node-levelup) in the constructor options that would be used to store data instead of coupling it to the filesystem.

**Comment by xaiki** [comment]
Wouldn't using levelup make this depend on binary (c++) modules ? These are a nightmare to build and distribute for apps like popcorn time.

How about local storage for the browser ?



—
Sent from Mobile,
I shalt be brief

**Comment by maxogden** [comment]
I think levelup is overkill because its a single json object. It should just be something super simple that implementers can write quickly, but async:

```js
{
  get: function (cb) {},
  put: function (data, cb) {}
}
```

**Comment by feross** [comment]
I think this belongs outside of bittorrent-dht. It should be really simple to persist and load the DHT:

```js
setInterval(function () {
  var data = dht.toArray()
  // get dht state every 15 minutes, and save it however you want
}, 15 * 60 * 1000)
```

On startup, just initialize the DHT with the saved table:

```js
var data = [] // get saved dht state however you want
var dht = new DHT({ bootstrap: data })
```

**Pull request closed without merging**